pyro/infer/mcmc/hmc.py

```diff
@@ -188,20 +215,108 @@ def _find_reasonable_step_size(self, z):
         return step_size
 
+    def _configure_adaptation(self, trace):
+        self._adapt_phase = True
+        self._adapt_window = self._adapt_initial_window
+
+        # from Stan, for small warmup_steps
+        if self._warmup_steps < 20:
+            self._adapt_start_buffer = self._warmup_steps
+        elif (self._adapt_start_buffer + self._adapt_end_buffer
+                + self._adapt_initial_window > self._warmup_steps):
+            self._adapt_start_buffer = int(0.15 * self._warmup_steps)
+            self._adapt_end_buffer = int(0.1 * self._warmup_steps)
+            self._adapt_window = (self._warmup_steps - self._adapt_start_buffer
+                                  - self._adapt_end_buffer)
+
+        # define ending pointer of the current window
+        self._adapt_window_ending = self._adapt_start_buffer
+        self._adapt_mass_matrix_phase_ending = self._warmup_steps - self._adapt_end_buffer
+
+        if self.adapt_step_size:
+            z = {name: node["value"] for name, node in self._iter_latent_nodes(trace)}
+            for name, transform in self.transforms.items():
+                z[name] = transform(z[name])
+            with pyro.validation_enabled(False):
+                self.step_size = self._find_reasonable_step_size(z)
+            self.num_steps = max(1, int(self.trajectory_length / self.step_size))
+            # make prox-center for Dual Averaging scheme
+            loc = math.log(10 * self.step_size)
+            self._step_size_adapt_scheme = DualAveraging(prox_center=loc)
+
+        if self.adapt_mass_matrix:
+            is_diag = not self.full_mass
+            self._mass_matrix_adapt_scheme = WelfordCovariance(diagonal=is_diag)
+
+    def _end_warmup(self):
+        self._adapt_phase = False
+        if self.adapt_step_size:
+            _, log_step_size_avg = self._step_size_adapt_scheme.get_state()
+            self.step_size = math.exp(log_step_size_avg)
+            self.num_steps = max(1, int(self.trajectory_length / self.step_size))
+
+    def _end_adapt_window(self):
+        if self._adapt_window_ending == self._warmup_steps:
+            self._end_warmup()
+            return
+
+        if self._adapt_window_ending == self._adapt_start_buffer:
+            if self.adapt_mass_matrix:
+                self._adapt_mass_matrix_phase = True
+        else:
+            if self.adapt_step_size:
+                self._step_size_adapt_scheme.prox_center = math.log(10 * self.step_size)
+                self._step_size_adapt_scheme.reset()
+
+            if self.adapt_mass_matrix:
+                self._inverse_mass_matrix = self._mass_matrix_adapt_scheme.get_covariance()
+                self._update_r_dist()
+                self._mass_matrix_adapt_scheme.reset()
+
+            self._adapt_window = 2 * self._adapt_window
+            if self._adapt_window_ending == self._adapt_mass_matrix_phase_ending:
+                self._adapt_mass_matrix_phase_ending = False
```

Comment:
It seems like this is wrongly being set to a boolean type here, or it should be `self._adapt_mass_matrix_phase`?

Reply:
Ah, nice catch, thanks!